Lib/typing.py

```diff
@@ -2412,10 +2414,72 @@ def utf8(value: bytes) -> bytes: ...
       def utf8(value: str) -> bytes: ...
       def utf8(value):
           # implementation goes here
+
+    Overload definition can be retrieved at runtime using the
+    get_overloads() function.
     """
+    try:
+        existing = get_overloads(func)
+    except AttributeError:
+        # Not a normal function; ignore.
+        pass
+    else:
+        if existing:
+            # If we are registering a variant with a lineno below or equal to that of the
+            # most recent existing variant, we're probably re-creating overloads for a
+            # function that already exists. In that case, we clear the existing variants
+            # to avoid leaking memory.
+            firstlineno = _get_firstlineno(func)
+            if firstlineno is not None:
+                existing_lineno = _get_firstlineno(existing[-1])
+                if existing_lineno is not None and firstlineno <= existing_lineno:
+                    clear_overloads(func)
+
+        key = _get_key_for_callable(func)
+        _overload_registry.setdefault(key, []).append(func)
     return _overload_dummy
 
 
+def _get_firstlineno(func):
+    # staticmethod, classmethod
+    func = getattr(func, "__func__", func)
+    if not hasattr(func, '__code__'):
+        return None
+    return func.__code__.co_firstlineno
+
+
+# {key: [overload]}
+_overload_registry = {}
+
+
+def get_overloads(func):
+    """Return all defined overloads for *func* as a sequence."""
+    key = _get_key_for_callable(func)
+    return _overload_registry.get(key, [])
+
+
+def clear_overloads(func=None):
+    """Clear all overloads for the given function (or all functions)."""
+    if func is None:
+        _overload_registry.clear()
+    else:
+        key = _get_key_for_callable(func)
+        _overload_registry.pop(key, None)
+
+
+def _get_key_for_callable(func):
+    """Return a key for the given callable.
+
+    This is used as a key in the overload registry.
+
+    If no key can be created (because the object is not of a supported type), raise
+    AttributeError.
```

Comment:
The AttributeError is a bit low-level. Could you catch it and turn it into a TypeError without cause or context, and with a specific message? That part shouldn't be performance-critical.

Reply:
Done